### PR TITLE
(CDAP-15787) 4 Aggregators

### PR DESCRIPTION
**Pull request by MEseifan**
Implemented FailureCollector into the following classes:

- DedupAggregator
- DistinctAggregator
- GroupByAggregator
- RowDenormalizerAggregator

### CODE REVIEW

File: core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/DedupAggregator.java [hunk 1]
```diff
@@ -144,21 +146,23 @@ private Schema getOutputSchema(Schema inputSchema) {
...
   private void validateSchema(Schema inputSchema, List<String> uniqueFields,
-                              @Nullable DedupConfig.DedupFunctionInfo function) {
+                              @Nullable DedupConfig.DedupFunctionInfo function, FailureCollector collector) {
     for (String uniqueField : uniqueFields) {
       Schema.Field field = inputSchema.getField(uniqueField);
       if (field == null) {
-        throw new IllegalArgumentException(String.format("Cannot do an unique on field '%s' because it does " +
-                                                           "not exist in inputSchema '%s'", uniqueField, inputSchema));
+        collector.addFailure(String.format("Cannot do an unique on field '%s' because it does " +
+                "not exist in inputSchema", uniqueField),
```
Comment (albertshau) [inline core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/DedupAggregator.java]: indentation (rest of PR too)

File: core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/DedupAggregator.java [hunk 2]
```diff
@@ -144,21 +146,23 @@ private Schema getOutputSchema(Schema inputSchema) {
...
 
   private void validateSchema(Schema inputSchema, List<String> uniqueFields,
-                              @Nullable DedupConfig.DedupFunctionInfo function) {
+                              @Nullable DedupConfig.DedupFunctionInfo function, FailureCollector collector) {
     for (String uniqueField : uniqueFields) {
       Schema.Field field = inputSchema.getField(uniqueField);
       if (field == null) {
-        throw new IllegalArgumentException(String.format("Cannot do an unique on field '%s' because it does " +
-                                                           "not exist in inputSchema '%s'", uniqueField, inputSchema));
+        collector.addFailure(String.format("Cannot do an unique on field '%s' because it does " +
```
Comment (albertshau) [inline core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/DedupAggregator.java]: This message is overly verbose, let's change it to:

Field '%s' does not exist in the input schema

File: core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/RowDenormalizerAggregator.java [hunk 3]
```diff
@@ -138,40 +140,49 @@ private Schema initializeOutputSchema() {
...
-      throw new IllegalArgumentException(
-        String.format("Keyfield '%s' does not exist in input schema %s", conf.getKeyField(), inputSchema));
-    } else if (inputSchema.getField(conf.getNameField()) == null) {
-      throw new IllegalArgumentException(
-        String.format("Namefield '%s' does not exist in input schema %s", conf.getNameField(), inputSchema));
-    } else if (inputSchema.getField(conf.getValueField()) == null) {
-      throw new IllegalArgumentException(
-        String.format("Valuefield '%s' does not exist in input schema %s", conf.getValueField(), inputSchema));
+      collector.addFailure(String.format("Keyfield '%s' does not exist in input schema", conf.getKeyField()),
+              "Provide a field that is matches a field from the input schema").withConfigProperty("keyField");
```
Comment (albertshau) [inline core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/RowDenormalizerAggregator.java]: don't need the corrective action here, I think it's pretty self explanatory what the user needs to do.

File: core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/RowDenormalizerAggregator.java [hunk 4]
```diff
@@ -138,40 +140,49 @@ private Schema initializeOutputSchema() {
...
     if (inputSchema.getField(conf.getKeyField()) == null) {
-      throw new IllegalArgumentException(
-        String.format("Keyfield '%s' does not exist in input schema %s", conf.getKeyField(), inputSchema));
-    } else if (inputSchema.getField(conf.getNameField()) == null) {
-      throw new IllegalArgumentException(
-        String.format("Namefield '%s' does not exist in input schema %s", conf.getNameField(), inputSchema));
-    } else if (inputSchema.getField(conf.getValueField()) == null) {
-      throw new IllegalArgumentException(
-        String.format("Valuefield '%s' does not exist in input schema %s", conf.getValueField(), inputSchema));
+      collector.addFailure(String.format("Keyfield '%s' does not exist in input schema", conf.getKeyField()),
```
Comment (albertshau) [inline core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/RowDenormalizerAggregator.java]: existing issue, but we should use plain english in the messages.

Key field '%s' does not exist in the input schema

File: core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/RowDenormalizerAggregator.java [hunk 5]
```diff
@@ -138,40 +140,49 @@ private Schema initializeOutputSchema() {
...
+              "Provide a field that is matches a field from the input schema").withConfigProperty("keyField");
+    } else {
+      Schema keyFieldSchema = inputSchema.getField(conf.getKeyField()).getSchema();
+
+      if (!((keyFieldSchema.isNullable() ? keyFieldSchema.getNonNullable().getType() : keyFieldSchema
+              .getType()).equals(Schema.Type.STRING))) {
+        collector.addFailure(String.format("Keyfield '%s' in the input record must be of type String or" +
```
Comment (albertshau) [inline core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/RowDenormalizerAggregator.java]: Key field '%s' in the input schema must be a string

File: core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/RowDenormalizerAggregator.java [hunk 6]
```diff
@@ -138,40 +140,49 @@ private Schema initializeOutputSchema() {
...
-        String.format("Namefield '%s' does not exist in input schema %s", conf.getNameField(), inputSchema));
-    } else if (inputSchema.getField(conf.getValueField()) == null) {
-      throw new IllegalArgumentException(
-        String.format("Valuefield '%s' does not exist in input schema %s", conf.getValueField(), inputSchema));
+      collector.addFailure(String.format("Keyfield '%s' does not exist in input schema", conf.getKeyField()),
+              "Provide a field that is matches a field from the input schema").withConfigProperty("keyField");
+    } else {
+      Schema keyFieldSchema = inputSchema.getField(conf.getKeyField()).getSchema();
+
+      if (!((keyFieldSchema.isNullable() ? keyFieldSchema.getNonNullable().getType() : keyFieldSchema
```
Comment (albertshau) [inline core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/RowDenormalizerAggregator.java]: it's hard to read `(!((`, can you break this up?

File: core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/RowDenormalizerAggregator.java (hunk 5 above)
Comment (albertshau) [reply core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/RowDenormalizerAggregator.java]: similar comments below for the other fields

File: core-plugins/src/test/java/io/cdap/plugin/batch/aggregator/RowDenormalizerConfigTest.java [hunk 7]
```diff
@@ -27,7 +29,7 @@
 /**
  * Test cases for {@link RowDenormalizerConfig}.
  */
-public class RowDenormalizerConfigTest {
+public class  RowDenormalizerConfigTest {
```
Comment (albertshau) [inline core-plugins/src/test/java/io/cdap/plugin/batch/aggregator/RowDenormalizerConfigTest.java]: remove extra space

File: core-plugins/widgets/GroupByAggregate-batchaggregator.json [hunk 8]
```diff
@@ -35,17 +35,17 @@
...
-              "First",
-              "Last",
-              "Max",
-              "Min",
-              "Stddev",
-              "Sum",
-              "Variance",
-              "CollectList",
-              "CollectSet"
+              "avg",
```
Comment (albertshau) [inline core-plugins/widgets/GroupByAggregate-batchaggregator.json]: these shouldn't change, pretty sure product wanted capitalized options in the UI.

**Review by albertshau — COMMENTED** [review]
also looks like there's unintended changes to google-cloud submodule hash

File: core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/DedupAggregator.java (hunk 1 above)
Comment (MEseifan) [reply core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/DedupAggregator.java]: Done

File: core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/DedupAggregator.java (hunk 2 above)
Comment (MEseifan) [reply core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/DedupAggregator.java]: Done

File: core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/RowDenormalizerAggregator.java (hunk 3 above)
Comment (MEseifan) [reply core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/RowDenormalizerAggregator.java]: Removed corrective action

File: core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/RowDenormalizerAggregator.java (hunk 4 above)
Comment (MEseifan) [reply core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/RowDenormalizerAggregator.java]: Done

Comment (MEseifan) [reply core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/RowDenormalizerAggregator.java]: Also changed all .withConfigProperty strings to constants in other PRs

File: core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/RowDenormalizerAggregator.java (hunk 5 above)
Comment (MEseifan) [reply core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/RowDenormalizerAggregator.java]: Done

File: core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/RowDenormalizerAggregator.java (hunk 6 above)
Comment (MEseifan) [reply core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/RowDenormalizerAggregator.java]: Done

File: core-plugins/src/test/java/io/cdap/plugin/batch/aggregator/RowDenormalizerConfigTest.java (hunk 7 above)
Comment (MEseifan) [reply core-plugins/src/test/java/io/cdap/plugin/batch/aggregator/RowDenormalizerConfigTest.java]: Done

File: core-plugins/widgets/GroupByAggregate-batchaggregator.json (hunk 8 above)
Comment (MEseifan) [reply core-plugins/widgets/GroupByAggregate-batchaggregator.json]: Reverted them to the capitalized version and implemented a map in the aggregator to handle the naming mismatch

**Review by albertshau — COMMENTED** [review]
indentation off with this one

**Comment by MEseifan** [comment]
Fixed indentation

**Comment by albertshau** [comment]
fixed indentation. Squash commits after fixing commit history

**Comment by albertshau** [comment]
https://builds.cask.co/browse/HYP-BAD592

**Review by albertshau — APPROVED** [review]

**Pull request merged**